internal/balancertesting/balancertesting.go

```diff
@@ -357,6 +357,14 @@ func (hc *FakeHealthChecker) New(_ context.Context, connection conn.Conn, tracke
 		if ch := hc.initialized[connection]; ch != nil {
 			close(ch)
 		}
+		context.AfterFunc(ctx, func() {
+			// Automatically force state to unhealthy after context is cancelled.
+			tracker := hc.updateHealthState(connection, health.StateUnhealthy, true)
+			if tracker == nil {
+				return
+			}
+			tracker.UpdateHealthState(connection, health.StateUnhealthy)
+		})
```

Comment:
This makes the test health checker context-aware, which reproduced the originally reported issue.